vault/resource_aws_secret_backend_role.go

```diff
@@ -130,25 +130,15 @@ func awsSecretBackendRoleWrite(d *schema.ResourceData, meta interface{}) error {
 	data := map[string]interface{}{
 		"credential_type": credentialType,
 	}
-	if policy != "" {
-		data["policy_document"] = policy
-	}
-	if len(policyARNs) != 0 {
-		data["policy_arns"] = policyARNs
-	}
-	if len(roleARNs) != 0 {
-		data["role_arns"] = roleARNs
-	}
+	data["policy_document"] = strings.TrimSpace(policy.(string))
+	data["policy_arns"] = policyARNs
+	data["role_arns"] = roleARNs
```

Comment:
Are these being added even if they're zero so that people will be able to set them to empty again? If so, what would you think about instead putting them behind a `GetOk` where, if OK, they get set?

Reply:
I feel like knowing that to remove these attributes if they were previously set, you need to set them to an empty list (or empty string, for `policy_document`) is counter intuitive. Two seemingly similar configurations have different results depending on the current state:

```tf
resource "" "" {
  policy_arns = []
}

# vs

resource "" "" {
  # nothing
}
```

If you apply these two from a fresh state, you get the same result. I'd expect if I applied this on another cluster that already has configuration would result in the same. If we use `GetOk` the non-explicit version will keep the old value.

Reply:
I like the first way of unsetting a field.

It's great for backwards compatibility. Since these 5 fields are all optional, some users will have omitted those fields from their configurations because they're using the present intentional or default settings. This could overwrite their previous settings.

It's great for forwards compatibility too. In Vault, [0 indicates the some TTLs should never expire](https://www.vaultproject.io/docs/concepts/tokens#token-time-to-live-periodic-tokens-and-explicit-max-ttls). If we ever changed STS TTLs to behave that way too, overwriting with 0 here would dramatically extend an STS token's expiration.